Data/ByteString/Builder/RealFloat.hs

```diff
@@ -0,0 +1,273 @@
+        m <- peek mOut
+        e <- peek eOut
+        return $ FD32 m e
+
+{-# INLINE ryu_d2s_fd #-}
+ryu_d2s_fd :: Double -> FloatingDecimal64
+ryu_d2s_fd f = unsafeDupablePerformIO $
+    alloca $ \mOut -> do
+      alloca $ \eOut -> do
+        c_ryu_d2s_fd (realToFrac f) mOut eOut
+        m <- peek mOut
+        e <- peek eOut
+        return $ FD64 m e
+
+asCBool :: Bool -> CUChar
+asCBool = fromIntegral . fromEnum
+
+{-# INLINE ryu_f2s_to_chars #-}
+ryu_f2s_to_chars :: Word32 -> Int32 -> Bool -> ByteString
+ryu_f2s_to_chars m e s = unsafeDupablePerformIO $ do
+    fp <- mallocByteString 16 :: IO (ForeignPtr Word8)
+    withForeignPtr fp $ \p ->
+      PS fp 0 . fromIntegral <$> c_ryu_f2s_to_chars (fromIntegral m) (fromIntegral e) (asCBool s) p
+
+{-# INLINE ryu_d2s_to_chars #-}
+ryu_d2s_to_chars :: Word64 -> Int32 -> Bool -> ByteString
+ryu_d2s_to_chars m e s = unsafeDupablePerformIO $ do
+    fp <- mallocByteString 25:: IO (ForeignPtr Word8)
+    withForeignPtr fp $ \p ->
+      PS fp 0 . fromIntegral <$> c_ryu_d2s_to_chars (fromIntegral m) (fromIntegral e) (asCBool s) p
+
+
+-- auxiliary fixed format printing functions
+decimalLength9:: Integral a => a -> Int
+decimalLength9 v
+  | v >= 100000000 = 9
+  | v >= 10000000 = 8
+  | v >= 1000000 = 7
+  | v >= 100000 = 6
+  | v >= 10000 = 5
+  | v >= 1000 = 4
+  | v >= 100 = 3
+  | v >= 10 = 2
+  | otherwise = 1
+
+decimalLength17 :: Integral a => a -> Int
+decimalLength17 v
+  | v >= 10000000000000000 = 17
+  | v >= 1000000000000000 = 16
+  | v >= 100000000000000 = 15
+  | v >= 10000000000000 = 14
+  | v >= 1000000000000 = 13
+  | v >= 100000000000 = 12
+  | v >= 10000000000 = 11
+  | v >= 1000000000 = 10
+  | v >= 100000000 = 9
+  | v >= 10000000 = 8
+  | v >= 1000000 = 7
+  | v >= 100000 = 6
+  | v >= 10000 = 5
+  | v >= 1000 = 4
+  | v >= 100 = 3
+  | v >= 10 = 2
+  | otherwise = 1
+
+sign :: RealFloat a => a -> Builder
+sign f = if f < 0 then char7 '-' else mempty
+
+specialStr :: RealFloat a => a -> Maybe Builder
+specialStr f
+  | isNaN f          = Just $ string7 "NaN"
+  | isInfinite f     = Just $ sign f `mappend` string7 "Infinity"
+  | isNegativeZero f = Just $ string7 "-0.0"
+  | f == 0           = Just $ string7 "0.0"
+  | otherwise        = Nothing
+
+-- show fixed floating point matching show / showFFloat output by dropping
+-- digits after exponentiation precision
+ryu_d2fixed :: Double -> Maybe Int -> Builder
+ryu_d2fixed f prec
+  | Just b <- specialStr f = b
+  | otherwise = sign f `mappend` showFixed m e' prec
+  where (FD64 m e) = ryu_d2s_fd f
+        olength = decimalLength17 m
+        -- NB: exponent in exponential format is e' - 1
+        e' = fromIntegral e + olength
+
+showFixed :: Word64 -> Int -> Maybe Int -> Builder
+showFixed m e prec =
+    case prec of
+       Nothing
+         | e <= 0 -> char7 '0'
+                  `mappend` char7 '.'
+                  `mappend` string7 (replicate (-e) '0')
+                  `mappend` foldr mappend mempty (toB ds)
```

Comment:
why not simply `mconcat`?

Reply:
Oversight on my end. https://github.com/haskell/bytestring/pull/222/commits/2202a4a3601103215779701045041d9ae8ba4bef